modules/qualitycontrol/QualityController.ts

```diff
@@ -231,6 +266,80 @@ export default class QualityController {
         return false;
     }
 
+    /**
+     * Adjusts codec, lastN or receive resolution based on the send resolution (of the outbound streams) and limitation
+     * reported by the browser in the WebRTC stats. Recovery is also attempted if the limitation goes away. No action
+     * is taken if the adaptive mode has been disabled through config.js.
+     *
+     * @param {ISourceStats} sourceStats - The outbound-rtp stats for a local video track.
+     * @returns {void}
+     */
+    _performQualityOptimizations(sourceStats: ISourceStats): void {
+        // Do not attempt run time adjustments if the adaptive mode is disabled.
+        if (!this._enableAdaptiveMode) {
+            return;
+        }
+
+        const { encodeResolution, localTrack, qualityLimitationReason, tpc } = sourceStats;
+        const trackId = localTrack.rtcId.toString();
+
+        if (encodeResolution === tpc.calculateExpectedSendResolution(localTrack)) {
+            if (this._limitedByCpuTimeout) {
+                window.clearTimeout(this._limitedByCpuTimeout);
+                this._limitedByCpuTimeout = undefined;
+            }
+
+            if (qualityLimitationReason === QualityLimitationReason.NONE
```

Comment:
If we move from being CPU limited to being bandwidth limited (not sure that's possible!) we are not doing anything, which I guess it's ok?

Reply:
If the bandwidth limitation kicks in after the cpu one gets cleared, I do not think its a good idea to increase lastN. Though the limitation is in upstream and increasing lastN affect downstream b/w, I still think its better to wait for b/w limitation to go away before ramping up lastN.

Reply:
Sounds good, just wanted to double check.